feat: (Platform) Select component creation

sKudum opened the pull request:
#### Please provide a link to the associated issue.
closes #2521

#### Please provide a brief summary of this pull request.

1. Select component creation in platform which supports datasource and  platform form.
2. Few accessibility changes addressed in **core** select as suggested by QA team for the purpose of testing.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [n/a ] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

sKudum commented:
netify url is not updated with latest changes, giving try without WIP in title

fkolar commented on libs/platform/src/lib/components/form/select/select/select.component.ts:
```diff
@@ -0,0 +1,321 @@
+    encapsulation: ViewEncapsulation.None,
+    providers: [{ provide: FormFieldControl, useExisting: SelectComponent, multi: true }]
+})
+export class SelectComponent extends BaseSelect implements OnInit, AfterViewInit, AfterViewChecked {
```
I am not sure if select shoudl have DataSource. DataSource are for components that requires to handle larger datasets. and for Select we had List binding

fkolar commented on apps/docs/src/app/platform/platform-documentation.routes.ts:
```diff
@@ -94,8 +94,8 @@ export const ROUTES: Routes = [
             {
                 path: 'select',
                 loadChildren: () =>
```
Please add example also for noSelectionValue.

fkolar commented on libs/platform/src/lib/components/form/select/commons/base-select.ts:
```diff
@@ -0,0 +1,759 @@
+     * and retrieve the best matching DataProvider that is set on application level
+     */
+    @Input()
+    entityClass: string;
```
EntityClass is only valid for datasources . We dont have usecase for DS yet for this component

fkolar commented on libs/platform/src/lib/components/form/select/commons/base-select.ts:
```diff
@@ -0,0 +1,759 @@
+import { FormField } from '../../form-field';
+import { TextAlignment } from '../../combobox';
+
+export type FdpSelectDataSource<T> = SelectDataSource<T> | Observable<T[]> | T[];
```
Please see my commend for datasource

fkolar commented on libs/platform/src/lib/components/form/select/commons/base-select.ts:
```diff
@@ -0,0 +1,759 @@
+}
+
+@Directive()
+export abstract class BaseSelect extends CollectionBaseInput implements AfterViewInit, OnDestroy {
```
CollectionBaseInput should have a list binding

fkolar commented on libs/platform/src/lib/components/form/select/select/select.component.ts:
```diff
@@ -0,0 +1,321 @@
+
+    /** Sets control aria-label attribute value */
+    @Input()
+    ariaLabel: string = null;
```
Not sure why you have several properties defined both in this class and parent class

sKudum commented:
Hi @Frank,

1).DataSource is removed used list binding of collectionBaseInput.
2).Removed unused code.
3).Added more examples including for no selection.
4).Add test cases testing more bindings with form-controls.


Few Points:
1).As per fiori spec select should support 2 columns, but no grouping(belongs to combobox) with flexbile column ratio- updated that feature as well.

2). Platform select is wrapper around core select and core select looks for fd-option component  as child element.by this not able to achieve template driven approach in platform. 

3).Select should support text clipping after 2/3 rows. did not find any disruption of it in  git book neither in fundamental-styles.- Are we planing to have to in feature once styles are ready?.

fkolar commented on libs/platform/src/lib/components/form/select/commons/base-select.ts:
```diff
@@ -0,0 +1,559 @@
+    /** Data for suggestion list */
+    @Input()
+    get list(): any {
+        return this._listData;
```
Have you checked the parent class for existing list Input?

fkolar reviewed (changes requested):
The rest looks good to me , we just need to reuse what is in parent classes

fkolar reviewed (approved):
Looks good

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:
```diff
@@ -0,0 +1,139 @@
+<fd-docs-section-title [id]="'mode'" [componentName]="'select'">
```
unnecessary binding

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:
```diff
@@ -0,0 +1,139 @@
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'semantic'" [componentName]="'select'">
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:
```diff
@@ -0,0 +1,139 @@
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'custom-select-template'" [componentName]="'select'">
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:
```diff
@@ -0,0 +1,139 @@
+<separator></separator>
+
+
+<fd-docs-section-title [id]="'select-mobile'" [componentName]="'select'">
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:
```diff
@@ -0,0 +1,139 @@
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'select-max-height'" [componentName]="'select'">
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:
```diff
@@ -0,0 +1,139 @@
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'select-none'" [componentName]="'select'">
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:
```diff
@@ -0,0 +1,139 @@
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'select-nowrap'" [componentName]="'select'">
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:
```diff
@@ -0,0 +1,139 @@
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'reactive-form'" [componentName]="'select'">
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-height/platform-select-max-height-example.component.html:
```diff
@@ -0,0 +1,8 @@
+<fdp-select 
+placeholder="Select an option"
+(selectionChange)="onSelect($event)"
+[maxHeight]="'250px'"
```
unnecessary binding

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-mode-example/platform-select-mode-example.component.ts:
```diff
@@ -0,0 +1,52 @@
+        return {
+            label: item.name + item.id,
+            value: item,
+            triggerValue: '( ' + item.id + ' )',
+            disabled: item.id === 'B' ? true : false,
```
```suggestion
            triggerValue: `(${item.id})`,
            disabled: item.id === 'B',
```

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-none/platform-select-none-example.component.html:
```diff
@@ -0,0 +1,8 @@
+placeholder="Select an option"
+width="7rem"
+(selectionChange)="onSelect($event)"
+[noValueLabel]="'None'"
```
unnecessary binding

Betrozov commented on libs/core/src/lib/select/option/option.component.ts:
```diff
@@ -24,6 +24,7 @@ import {
 
 import { KeyUtil } from '../../utils/functions/key-util';
 import { OptionsInterface } from '../options.interface';
+import { By } from '@angular/platform-browser';
```
do you need it?

Betrozov commented on libs/platform/src/lib/components/form/select/select/select.component.html:
```diff
@@ -0,0 +1,76 @@
+[attr.aria-labelledby]="value"
+[attr.tooltip]="value"
+[attr.aria-readonly]="readonly"
+[attr.role]="'listbox'"
```
unnecessary binding

Betrozov commented on libs/platform/src/lib/components/form/select/select/select.component.ts:
```diff
@@ -0,0 +1,254 @@
+
+    /** @hidden
+     */
+    ngOnInit(): void {}
```
please remove empty state

sKudum replied:
removal of it leads to compile time error as =>"at SelectComponent.push.FWTC.BaseInput.ngOnInit"

Betrozov commented on libs/platform/src/lib/components/form/select/select/select.component.ts:
```diff
@@ -0,0 +1,254 @@
+            }
+            // by default text will be overlapped, below it will help to truncate
+            if (this.noWrapText) {
+                optionItem.querySelector('.fd-list__title').setAttribute(
+                    'title', optionItem.querySelector('.fd-list__title').innerHTML);
+                optionItem.querySelector('.fd-list__title').setAttribute(
+                    'aria-label', optionItem.querySelector('.fd-list__title').innerHTML);
+                optionItem.querySelector('.fd-list__title').classList.add('fd-list__title--no-wrap');
```
```suggestion
                const listTitle = optionItem.querySelector('.fd-list__title');
                const listTitleInnerHTML = optionItem.querySelector('.fd-list__title').innerHTML;
                listTitle.setAttribute('title', listTitleInnerHTML);
                listTitle.setAttribute('aria-label', listTitleInnerHTML);
                listTitle.classList.add('fd-list__title--no-wrap');
```

Betrozov commented on libs/platform/src/lib/components/form/select/select/select.component.ts:
```diff
@@ -0,0 +1,254 @@
+     * @hidden
+     */
+    _isSelectedOptionItemByDisplayValue(selectedItem: any): boolean {
+        return this.selected && this.selected.label === selectedItem;
```
```suggestion
        return this.selected?.label === selectedItem;
```

sKudum replied:
> 
Can't do this change as "Expression expected error "was getting displayed on compilation

Betrozov commented on libs/platform/src/lib/components/form/select/commons/base-select.ts:
```diff
@@ -0,0 +1,551 @@
+      * @hidden
+      * Used to define if contentDensity value is 'compact' or not.
+    */
+   _isCompact = this._contentDensity === 'compact';
```
seems it not used

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-semantic-state-example/platform-select-semantic-state-example.component.ts:
```diff
@@ -0,0 +1,45 @@
+            label: item.name + item.id,
+            value: item,
+            triggerValue: `(${item.id})`,
+            disabled: item.id === 'B' ? true : false,
```
```suggestion
            disabled: item.id === 'B',
```

Betrozov commented on libs/platform/src/lib/components/form/select/select/select.component.ts:
```diff
@@ -0,0 +1,255 @@
+            // by default text will be overlapped, below it will help to truncate
+            if (this.noWrapText) {
+                const listTitle = optionItem.querySelector('.fd-list__title');
+                const listTitleInnerHTML = optionItem.querySelector('.fd-list__title').innerHTML;
```
```suggestion
                const listTitleInnerHTML = listTitle.innerHTML;
```

puru-hk commented:
In RTL mode UI is not proper attache screenshot
![Uploading image.png…]()

puru-hk commented:
In RTL Mobile display is not proper:

![image](https://user-images.githubusercontent.com/32538291/108956330-847e3380-7695-11eb-84fc-4af27f30905d.png)

sKudum commented:
> In RTL Mobile display is not proper:
> 
> ![image](https://user-images.githubusercontent.com/32538291/108956330-847e3380-7695-11eb-84fc-4af27f30905d.png)
> 
> ![image](https://user-images.githubusercontent.com/32538291/108956493-ca3afc00-7695-11eb-86b6-c7defb73da97.png)

Hi @puru-hk ,

Issues are fixed , please have a look.

After fix:

![2021-02-24_15-16-25 (1)](https://user-images.githubusercontent.com/53487468/108982279-9ff83700-76b3-11eb-80b5-1b6d3bb06597.gif)

puru-hk commented:
Looks good

JKMarkowski reviewed (approved):
LGTM - Great job with fixing tab out issue
